thegraph/src/subscriptions/ticket.rs

```diff
@@ -0,0 +1,416 @@
+//! A subscription ticket is an authorization token that allows a user to send queries to the graph network.
+//!
+//! # Ticket
+//!
+//! A ticket has 2 parts:
+//!
+//!  1. Payload
+//!  2. Signature
+//!
+//! The signature is always the last 65 bytes of the ticket. The ticket should be Base64Url encoded when they are sent
+//! to gateways along with queries.
+//!
+//! ## Payload
+//!
+//! The ticket payload must be a [CBOR](https://www.rfc-editor.org/rfc/rfc7049)-encoded map and MUST containing the
+//! following ticket claims:
+//!
+//!  1. *chain_id:* Chain ID (EIP-155) for the chain on which the subscriptions contract is deployed.
+//!  2. *contract:* Address of the subscriptions contract.
+//!  3. *signer:* Signer address that is authorized to sign the ticket. This address should be the user's address or
+//!               one of the user's authorized signers.
+//!  4. *user (optional):* User address associated with the subscription. Required to when the authorized `signer` is
+//!                        not the `user` associated with a subscription. When omitted, the `signer` is implied to be
+//!                        equal to the `user`.
+//!
+//! Other optional fields may be supported at the gateway operator's discretion. See [`TicketClaims`] for other
+//! supported fields.
+//!
+//! Note that the gateway address is implied to be the owner of the subscriptions contract. In the future, we may need
+//! to explicitly identify the intended recipient gateway to prevent attacks where a gateway proxies requests.
+//!
+//! ## Signature
+//!
+//! Signing and verification of tickets uses an Ethereum signed message ([EIP-191](https://eips.ethereum.org/EIPS/eip-191),
+//! `personal_sign`) constructed from the ticket claims.
+//!  
+//!  - The message MUST be UTF-8 encoded.
+//!  - Fields MUST be ordered lexicographically by field name.
+//!  - Each field MUST be immediately followed by an ASCII LF character (`0x0a`).
+//!  - Each field name and value MUST be separated by `": "`.
+//!  - Any byte string value MUST be formatted as `0x` followed by its hex-encoded bytes.
+//!
+//! See [`TicketClaims::to_verification_message`] method for the implementation details.
+
+use std::io::{Cursor, Write as _};
+
+pub use alloy_chains::{Chain, NamedChain};
+use base64::{prelude::BASE64_URL_SAFE_NO_PAD, Engine as _};
+use ethers::signers::Wallet;
+use ethers_core::k256::ecdsa::SigningKey;
+use ethers_core::types::Signature;
+
+use crate::types::Address;
+
+/// Claims that are encoded in a ticket.
+#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
+pub struct TicketClaims {
```

Comment:
Now's probably a good time to rename "ticket" to "query key" once and for all.

Reply:
I'm also not sure about calling this "claims", since the only thing being claimed really is the signer. Most of these fields are settings that the creator is requesting the gateway to enforce.

Reply:
I believe that using the "Token" and "Claims" terminology improves the DX (Developer experience) as one can do mental parallelism with the _JSON Web Token_ spec for web services authorization and authentication.

-----

> Now's probably a good time to rename "ticket" to "query key" once and for all.

This "ticket" is a token associated with a subscription (free tier, paid, etc.). It is used by the different services for authentication/authorization.

I believe that we should follow the very much extended/well-known terminology of the JSON Web Token and rename it into "token" (e.g., `AuthTokenClaims`).

> I'm also not sure about calling this "claims", since the only thing being claimed really is the signer. Most of these fields are settings that the creator is requesting the gateway to enforce.

Following your reasoning, I kind of agree 🫤

However, I believe that "claims" is the least bad name for the token payload because it aligns with the [JWT Claims](https://auth0.com/docs/secure/tokens/json-web-tokens/json-web-token-claims).

Reply:
Ok, fair enough. I'm fine with "ticket" -> "auth token" (just "token" feels a bit too overloaded in Web3), and "ticket payolad" -> "auth token claims". @cmwhited any strong opinions on this?